### PR TITLE
Added contributing.md file.

### PR DESCRIPTION
**Pull request by JanEricNitschke**
Added a first draft of a contributing file. Please have a look for types and other improvements.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,89 @@
+# Contributing to awpy
+
+Hi! We're happy to have you here. Thank you in advance for your contribution to awpy.
+
+## Python code
+
+To make sure that awpy code stays readeable and works correctly we are making use of
```
Comment (pnxenopoulos) [inline CONTRIBUTING.md:7]: To make sure that awpy code stays readable and works correct, we are making use of a variety of helpful tools.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+a variaty of helpfull tools.
+
+At the time of writing these are:
```
Comment (pnxenopoulos) [inline CONTRIBUTING.md:10]: You can just say "We use the following tools:" since this is a living document.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+- [black](https://github.com/psf/black): An uncompromising Python code formatter.
+- [ruff](https://github.com/astral-sh/ruff): An extremely fast Python linter, written in Rust.
+- [pyright](https://github.com/microsoft/pyright): A static type checker for Python.
+- [pylint](https://github.com/pylint-dev/pylint): A static code analyser for Python.
+- [pytest](https://docs.pytest.org/en/7.4.x/): A mature full-featured Python testing tool.
+- [pre-commit](https://pre-commit.com/): A framework for managing and maintaining multi-language pre-commit hooks.
+
+To make sure that you pull request is easy to review and can be merged quickly please
```
Comment (pnxenopoulos) [inline CONTRIBUTING.md:19]: Please install these tools before you open a pull request to ensure that your contributions are easy to review.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+please install these after cloning the repo.
+
+For that simply run the following commands:
```
Comment (pnxenopoulos) [inline CONTRIBUTING.md:22]: To install the aforementioned tools, simply run the following:

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+```shell
+pip install -r tests/requirements.txt
+pre-commit install --install-hooks
+```
+
+After you have made your changes locally run these tools to make sure there are no
```
Comment (pnxenopoulos) [inline CONTRIBUTING.md:28]: After you have made your changes locally, use these tools to surface bugs or code smells by running the following:

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+bugs or code smells have snook in. This can be done by running:
+
+```shell
+pre-commit run --all-files --show-diff-on-failure  # black formatting, ruff linting
+pyright  # type checking
+pylint awpy  # more linting
+coverage run -m pytest --durations=10  # python tests
+coverage report -m
```
Comment (pnxenopoulos) [inline CONTRIBUTING.md:36]: Can add a comment here like `# produces text-based coverage report`

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+```
+
+The test part can take a while as it has to download a handful of demos to make sure
```
Comment (pnxenopoulos) [inline CONTRIBUTING.md:39]: The `coverage run -m pytest --durations=10` command, which runs the Python tests, can take a while as it has to not only download a handful of demos but also parse them. These tests must pass for a pull request to be merged into the main branch.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+that the parser is working correctly for all kinds of platforms and edge cases.
+
+If you do not have the time or bandwidth you can omit this part if absolutely necessary.
```
Comment (pnxenopoulos) [inline CONTRIBUTING.md:42]: If you do not have the time or bandwidth, you can omit running the test command locally, since Github actions will run them, as well.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+The continous integration setup on GitHub will then run these for you.
+
+### Testing/Coverage
+
+If you are fixing a bug or adding a new feature make sure to also add [unit tests](https://en.wikipedia.org/wiki/Unit_testing)
+that cover the original bug or your new functionality. If you are new to that have a
```
Comment (pnxenopoulos) [inline CONTRIBUTING.md:48]: If you are new to writing unit tests, look at the aforementioned link, or check out the tests directory in awpy to see how our existing tests are built.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+look at the pytest documentation linked above. You can also check out the [tests](tests)
+directory in awpy to see how the existing code is being tested.
+
+## Go code
+
+If you are making changes or additions to the Go code then a look at and run
+[golangci-lint](https://github.com/golangci/golangci-lint).
+
+To install it simply follow the [instructions](https://golangci-lint.run/usage/install/#local-installation) on their website.
```
Comment (pnxenopoulos) [inline CONTRIBUTING.md:57]: To install it, simply ...

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+To run it do the following:
+
+```
+cd awpy/parser  # Change into the directory containing the go files
+golangci-lint run  # Run the linters
+```
+
+Additionally run the tests of the go code via:
```
Comment (pnxenopoulos) [inline CONTRIBUTING.md:65]: Additionally, run the Go code tests with:

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+```
+go test -covermode=count -coverprofile=coverage.out
+go tool cover -html=coverage.out -o coverage.html
+go test -fuzz=FuzzConvertRank -fuzztime 30s
+go test -fuzz=FuzzDetermineSecond -fuzztime 30s
+go test -fuzz=FuzzParseTeamBuy -fuzztime 30s
+```
+
+Do not forget to also run the python tests as explained above.
```
Comment (pnxenopoulos) [inline CONTRIBUTING.md:74]: Don't forget to run the Python tests as explained above -- changes in the Go code likely change the JSON output!

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+If you are adding functionality aim to also add go tests to avoid bug and future regressions.
```
Comment (pnxenopoulos) [inline CONTRIBUTING.md:76]: If you are adding functionality with your contribution, try to add Go tests to avoid bugs and future regressions. Incase changes to the Go code were required, make sure it does not crash on certain kinds of demos. If you have a new demo, please let xeno know and he can add it to the data repository used for the tests.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+In case changes to the go code were required to make sure it does not crash on certain
+kinds of demos please also add unit tests for these demos in the python tests.
+
+For us to run these tests ourselves to verify the results and make sure no future changes
```
Comment (pnxenopoulos) [inline CONTRIBUTING.md:80]: If you are adding a test that requires a specific demo, please let us know so that we can include them in our set of testing demos. While developing, you can also edit the `test_data.json` file to include wherever you are hosting the test demo file.

**Review by pnxenopoulos — CHANGES_REQUESTED** [review]
I left some changes, mostly grammatical. Otherwise, this is a great addition, thanks once again!

**Comment by JanEricNitschke** [comment]
> I left some changes, mostly grammatical. Otherwise, this is a great addition, thanks once again!

Should hopefully all be addressed now. This has never been my strong suit. Neither in german nor in english.
